Sources/Build/BuildPlan.swift

```diff
@@ -259,18 +259,64 @@ public final class ProductBuildDescription {
     }
 
     /// The objects in this product.
-    let objects: [AbsolutePath]
+    // Computed during build planning.
+    fileprivate(set) var objects: [AbsolutePath] = []
 
-    /// Any addition flags to be added. These flags are expected to be computed during build planning.
+    /// The dynamic libraries this product needs to link with.
+    // Computed during build planning.
+    fileprivate(set) var dylibs: [ProductBuildDescription] = []
+
+    /// Any additional flags to be added. These flags are expected to be computed during build planning.
     fileprivate var additionalFlags: [String] = []
 
     /// Create a build description for a product.
-    init(product: ResolvedProduct, objects: [AbsolutePath], buildParameters: BuildParameters) {
+    init(product: ResolvedProduct, buildParameters: BuildParameters) {
+        assert(product.type != .library(.none), "None type libraries should not be described.")
         self.product = product
         self.buildParameters = buildParameters
-        self.objects = objects
     }
 
+    /// Computes (lazily) and returns the target and product dependencies.
+    ///
+    /// This property returns three things:
+    ///     - staticTargets: The targets which can be linked statically in this product.
+    ///     - dylibs: The products which needs to be dynamically linked.
+    ///     - systemModules: The system module targets this product depends on.
+    public lazy var dependencies: (staticTargets: [ResolvedModule], dylibs: [ResolvedProduct], systemModules: [ResolvedModule]) = {
```

Comment:
FWIW, I think lazy is a little scary for this for two reasons: 1. it isn't thread safe so if we move to doing some part of build planing in parallel it could be dangerous, 2. isn't this something which should be done explicitly at one point in the build, errors diagnosed, and then kept forever?

Not something that needs to change in this PR, though...